### PR TITLE
Use new geckoview-omni package, which includes Glean directly

### PR DESCRIPTION
**Pull request by badboy**
The native Glean core code is now shipped as part of GeckoView, within
the `-omni` packages.
The Glean Gradle plugin takes care that this all fits together nicely.
Downstream users will need to depend on the `glean-native` package for
tests now.

For now Android Components still needs to declare an explicit dependency
on `org.mozilla.telemetry:glean` as well.
service-glean needs to know what to reexport and some other components
depend on it directly to be able to use it. These might not depend on
geckoview, so have no other way to get that dependency.

Now here's the catch: Because GeckoView defines `org.mozilla.telemetry:glean` as a dependency,
consumers that transitively depend on GeckoView (through
`browser-engine-gecko`) will get the latest version of `org.mozilla.telemetry:glean` through it.
Gradle unifies the versions.
It would be problematic if a-c depends on a higher version, but that
should not happen (though we don't have any technical limitations on it
yet).
If Glean does a breaking change release and gets that into
mozilla-central, the version in a-c needs to be manually updated too.
For other releases it can be left untouched.

---

cc @agi for review as well.

I tested this configuration locally in Fenix already.
Still we should keep an eye on when this lands in Fenix to validate that everything works as expected.
Our automated tests should catch if anything is really off, but this is a big change for how we ship Glean so worth a close look.

### CODE REVIEW

File: buildSrc/src/main/java/Dependencies.kt
Comment (agi) [inline buildSrc/src/main/java/Dependencies.kt:34]: Do we need this? I thought the intention was that AC wouldn't depend on glean directly anymore

**Review by Dexterp37 — APPROVED** [review]
Thank you for all the great work culminating in this succint PR. 💯

**Comment by badboy** [comment]
putting back into draft mode while I figure out some details

File: buildSrc/src/main/java/Dependencies.kt
Comment (badboy) [reply buildSrc/src/main/java/Dependencies.kt:34]: Unfortunately yes. I extended the commit message, also visible in the PR description above now, with some reasoning.

Note: we explicitly called this out in the doc back then as an optional step to try. I will file a followup bug to figure out how we can improve that.

**Comment by agi** [comment]
> For now Android Components still needs to declare an explicit dependency
on org.mozilla.telemetry:glean as well.
service-glean needs to know what to reexport and some other components
depend on it directly to be able to use it. These might not depend on
geckoview, so have no other way to get that dependency.

I don't understand this part, can you elaborate? is it because some components don't depend on GeckoView but depend on glean?

> Now here's the catch: Because GeckoView defines org.mozilla.telemetry:glean as a dependency,
consumers that transitively depend on GeckoView (through
browser-engine-gecko) will get the latest version of org.mozilla.telemetry:glean through it.
Gradle unifies the versions.

I think we need to figure this out before this lands, I think having the gradle plugin apply a ResolutionStrategy for glean should be enough? https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html

**Comment by badboy** [comment]
> > For now Android Components still needs to declare an explicit dependency
> > on org.mozilla.telemetry:glean as well.
> > service-glean needs to know what to reexport and some other components
> > depend on it directly to be able to use it. These might not depend on
> > geckoview, so have no other way to get that dependency.
> 
> I don't understand this part, can you elaborate? is it because some components don't depend on GeckoView but depend on glean?

Yup.
`service-glean` is a small shim around Glean: https://github.com/mozilla-mobile/android-components/blob/2fb230cc7c11afa942171245060ce3ee275acb21/components/service/glean/build.gradle#L50
It is what e.g. Fenix uses then.
Others within a-c then depend on that, e.g. [here in sync-telemetry](https://github.com/mozilla-mobile/android-components/blob/2fb230cc7c11afa942171245060ce3ee275acb21/components/support/sync-telemetry/build.gradle#L45). 
These don't depend on GeckoView, so somewhere along the line the Glean version needs to be known.

> 
> > Now here's the catch: Because GeckoView defines org.mozilla.telemetry:glean as a dependency,
> > consumers that transitively depend on GeckoView (through
> > browser-engine-gecko) will get the latest version of org.mozilla.telemetry:glean through it.
> > Gradle unifies the versions.
> 
> I think we need to figure this out before this lands, I think having the gradle plugin apply a ResolutionStrategy for glean should be enough? [docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html](https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html)

So the Resoution stragegy could fail if it detects something that shouldn't happen?
I can look into that tomorrow.

**Comment by agi** [comment]
> > I think we need to figure this out before this lands, I think having the gradle plugin apply a ResolutionStrategy for glean should be enough? [docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html](https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html)
> 
> So the Resoution stragegy could fail if it detects something that shouldn't happen?
> I can look into that tomorrow.

Yes something like `failOnVersionConflict` but just for glean, I don't know if there's something ready for it or if we need to write some custom code for it (I'm totally available to help you if it turns out it's tricky to do).

**Comment by badboy** [comment]
> > > I think we need to figure this out before this lands, I think having the gradle plugin apply a ResolutionStrategy for glean should be enough? [docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html](https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html)
> > 
> > 
> > So the Resoution stragegy could fail if it detects something that shouldn't happen?
> > I can look into that tomorrow.
> 
> Yes something like `failOnVersionConflict` but just for glean, I don't know if there's something ready for it or if we need to write some custom code for it (I'm totally available to help you if it turns out it's tricky to do).

The fenix patch now has the temporary workaround: [`6cec80b` (#20889)](https://github.com/mozilla-mobile/fenix/pull/20889/commits/6cec80b43d65dc7ab5461de5f6a037bf2c584d8f)

We'll get that fix into Glean proper ([bug 1726765](https://bugzilla.mozilla.org/show_bug.cgi?id=1726765)), but both the a-c patch and the fenix patch can already land.

**Review by agi — APPROVED** [review]
With the conflict resolution strategy in Fenix/the glean plugin this looks good to me. I'm not actually a peer on AC so this will also need to be approved by @csadilek and @pocmo

**Review by csadilek — CHANGES_REQUESTED** [review]
Thank you @badboy for moving this forward! So close now!

I am just marking as "request changes" until we figure out: https://github.com/mozilla-mobile/fenix/pull/20889#discussion_r694322076

I wanted to create a release build, compare APK sizes (just to double-check) and ask QA for a quick round of smoke testing before we land this all. Let's do that once we worked through the issue above.

**Comment by badboy** [comment]
Waiting on m-c changes to land before this works.

**Comment by badboy** [comment]
https://github.com/mozilla-mobile/android-components/pull/10923 landed a geckoview release that has all the necessary things (in the `-omni` variant), so this should now be working.

**Review by csadilek — APPROVED** [review]
@badboy OK, I created a local Nightly build and smoke tested with this change. All looks good to me now!

APK size is ~150K smaller than before.
Pings/events are sent correctly.

**Comment by csadilek** [comment]
Closing/Reopening to restart the builds that timed out.

**Comment by csadilek** [comment]
Hm TC doesn't want to build this. @bhearsum would you have any ideas? Looks like `[taskcluster:error] Task has been aborted prematurely. Reason: worker-shutdown` is happening here twice in a row: https://firefox-ci-tc.services.mozilla.com/tasks/foo_8kJnRpC2GfWW8Cov2Q/runs/5/logs/public/logs/live.log

**Comment by csadilek** [comment]
Ah looks like the last CI run pointed to a valid problem again:

```
[task 2021-09-07T23:53:12.974Z] * What went wrong:
[task 2021-09-07T23:53:12.974Z] Execution failed for task ':browser-storage-sync:lintAnalyzeRelease'.
[task 2021-09-07T23:53:12.974Z] > Could not resolve all files for configuration ':browser-storage-sync:releaseUnitTestCompileClasspath'.
[task 2021-09-07T23:53:12.974Z]    > Could not find org.mozilla.telemetry:glean-forUnitTests:40.1.0.
[task 2021-09-07T23:53:12.974Z]      Required by:
[task 2021-09-07T23:53:12.974Z]          project :browser-storage-sync
```

**Comment by csadilek** [comment]
@badboy OK, looks like we'll need to update this line https://github.com/mozilla-mobile/android-components/blob/main/buildSrc/src/main/java/Dependencies.kt#L154 to use `"org.mozilla.telemetry:glean-native-forUnitTests:${Versions.mozilla_glean}"`. It's used here: https://github.com/mozilla-mobile/android-components/blob/main/components/browser/storage-sync/build.gradle#L55

**Comment by bhearsum** [comment]
> Hm TC doesn't want to build this. @bhearsum would you have any ideas? Looks like `[taskcluster:error] Task has been aborted prematurely. Reason: worker-shutdown` is happening here twice in a row: https://firefox-ci-tc.services.mozilla.com/tasks/foo_8kJnRpC2GfWW8Cov2Q/runs/5/logs/public/logs/live.log

Sorry for the slow reply - this typically happens if a spot instance shuts down mid-build. It looks like you've managed to get past it?

**Comment by csadilek** [comment]
> Sorry for the slow reply - this typically happens if a spot instance shuts down mid-build. It looks like you've managed to get past it?

@bhearsum No problem. We got passed the build not running i.e., we're seeing the build fail now, but in the toolchain task above you see the build failed in 12m, but the task/check completed in 104m and isn't marked as failure. 🤔

**Comment by bhearsum** [comment]
> > Sorry for the slow reply - this typically happens if a spot instance shuts down mid-build. It looks like you've managed to get past it?
> 
> @bhearsum No problem. We got passed the build not running i.e., we're seeing the build fail now, but in the toolchain task above you see the build failed in 12m, but the task/check completed in 104m and isn't marked as failure. thinking

It looks like task `exception` states may not be reported back to GitHub as failures (although I'm honestly not sure....). Close/re-open to try again is the best solution we have right now for intermittent failures.

**Comment by badboy** [comment]
> @badboy OK, looks like we'll need to update this line [https://github.com/mozilla-mobile/android-components/blob/main/buildSrc/src/main/java/Dependencies.kt#L154](https://github.com/mozilla-mobile/android-components/blob/main/buildSrc/src/main/java/Dependencies.kt?rgh-link-date=2021-09-08T02%3A05%3A26Z#L154) to use `"org.mozilla.telemetry:glean-native-forUnitTests:${Versions.mozilla_glean}"`. It's used here: [https://github.com/mozilla-mobile/android-components/blob/main/components/browser/storage-sync/build.gradle#L55](https://github.com/mozilla-mobile/android-components/blob/main/components/browser/storage-sync/build.gradle?rgh-link-date=2021-09-08T02%3A05%3A26Z#L55)

Thanks. Not sure how I missed that before.

**Comment by badboy** [comment]
> [task 2021-09-13T10:25:43.675Z]             > Could not GET 'https://plugins.gradle.org/m2/org/jetbrains/kotlin/kotlin-util-io/1.4.31/kotlin-util-io-1.4.31.pom'. Received status code 500 from server: Internal Server Error

D'oh!

**Comment by csadilek** [comment]
Alright, this is green! Let's coordinate landing together with https://github.com/mozilla-mobile/fenix/pull/20889 tomorrow morning.

**Comment by badboy** [comment]
> Alright, this is green! Let's coordinate landing together with [mozilla-mobile/fenix#20889](https://github.com/mozilla-mobile/fenix/pull/20889) tomorrow morning.

Who's morning? (Just so I know when to look out for it)

**Comment by csadilek** [comment]
> Who's morning? (Just so I know when to look out for it)

Had to planned to do it in mine (EST), but let me know if you want to take care of it. Important part is to add the A-C version upgrade to your PR, or your changes to the A-C version upgrade.

**Comment by badboy** [comment]
> > Who's morning? (Just so I know when to look out for it)
> 
> Had to planned to do it in mine (EST), but let me know if you want to take care of it. Important part is to add the A-C version upgrade to your PR, or your changes to the A-C version upgrade.

Nope, please go ahead during a time convenient for you.

**Pull request merged**
